Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

```diff
@@ -0,0 +1,312 @@
+  Over.homMk (snd f g)
+
+theorem snd'_left : (snd' f g).left = snd f g := rfl
+
+variable {f g}
+
+@[reassoc]
+theorem condition : fst f g ≫ f = snd f g ≫ g :=
+  Over.w (fst' f g)
+
+variable (f g) in
+@[ext]
+theorem hom_ext {W : C} {φ₁ φ₂ : W ⟶ pullbackObj f g} (h₁ : φ₁ ≫ fst _ _ = φ₂ ≫ fst _ _)
+    (h₂ : φ₁ ≫ snd _ _ = φ₂ ≫ snd _ _) :
+    φ₁ = φ₂ := by
+  let adj := mapPullbackAdj g
+  let U : Over Z := Over.mk (φ₁ ≫ snd f g)
+  let φ₁' : U ⟶ (pullback g).obj (Over.mk f) := Over.homMk φ₁
+  let φ₂' : U ⟶ (pullback g).obj (Over.mk f) := Over.homMk φ₂ (by simpa using h₂.symm)
+  have : φ₁' = φ₂' := by
+    apply (adj.homEquiv U _).symm.injective
+    apply (forget X).map_injective
+    simpa using h₁
+  exact congr_arg CommaMorphism.left this
+
+section Lift
+
+variable {W : C} (a : W ⟶ Y) (b : W ⟶ Z) (h : a ≫ f = b ≫ g := by cat_disch)
+
+/-- Given morphisms `a : W ⟶ Y` and `b : W ⟶ Z` satisfying `a ≫ f = b ≫ g`,
+constructs the unique morphism `W ⟶ pullbackObj f g` which lifts `a` and `b`. -/
+def lift : W ⟶ pullbackObj f g :=
+  (((mapPullbackAdj g).homEquiv (Over.mk b) (Over.mk f)) (Over.homMk a)).left
+
+@[reassoc (attr := simp)]
+theorem lift_fst : lift a b h ≫ fst f g = a := by
+  let adj := mapPullbackAdj g
+  let a' : (Over.map g).obj (Over.mk b) ⟶ Over.mk f := Over.homMk a h
+  have : (Over.map g).map (adj.homEquiv (.mk b) (.mk f) (Over.homMk a)) ≫ fst' f g = a' := by
+    simp only [Functor.id_obj, ← Adjunction.homEquiv_counit, Equiv.symm_apply_apply, adj, a']
+  exact congr_arg CommaMorphism.left this
+
+@[reassoc (attr := simp)]
+theorem lift_snd : lift a b h ≫ snd f g = b := by
+  simp [lift]
+
+end Lift
+
+section PullbackMap
+
+variable (f g)
+
+/-- The functoriality of `pullbackObj f g` in both arguments: Given a map from the pullback cospans
+of `f' : Y' ⟶ X'` and `g' : Z' ⟶ X'` to the pullback cospan of `f : Y ⟶ X` and `g : Z ⟶ X`
+as in the diagram below
+```
+Y' ⟶ Y
+  ↘   ↘
+  X' ⟶ X
+  ↗   ↗
+Z' ⟶ Z
+```
+if the morphisms `g'` and `g` both have chosen pullbacks, then we get an induced morphism
+`pullbackMap f g f' g' comm₁ comm₂` from the chosen pullback of
+`f' : Y' ⟶ X'` along `g'` to the chosen pullback of `f : Y ⟶ X` along `g`.
+Here `comm₁` and `comm₂` are the commutativity conditions of the squares in the diagram above.
+-/
+def pullbackMap {Y' Z' X' : C} (f' : Y' ⟶ X') (g' : Z' ⟶ X') [ChosenPullback g']
+    (γ₁ : Y' ⟶ Y) (γ₂ : Z' ⟶ Z) (γ₃ : X' ⟶ X)
+    (comm₁ : f' ≫ γ₃ = γ₁ ≫ f := by cat_disch) (comm₂ : g' ≫ γ₃ = γ₂ ≫ g := by cat_disch) :
+    pullbackObj f' g' ⟶ pullbackObj f g :=
+  lift (fst f' g' ≫ γ₁) (snd f' g' ≫ γ₂)
+    (by rw [assoc, ← comm₁, ← assoc, condition, assoc, comm₂, assoc])
+
+/-- The canonical morphism `pullbackObj f g ⟶ pullbackObj (f ≫ i) (g ≫ i)`. -/
+def pullbackMapDesc {X' : C} (i : X ⟶ X') [ChosenPullback (g ≫ i)] :
+    pullbackObj f g ⟶ pullbackObj (f ≫ i) (g ≫ i) :=
+  pullbackMap (f ≫ i) (g ≫ i) f g (𝟙 _) (𝟙 _) i
```

Comment:
I would suggest removing this as it is no longer very much required for the definitions below.